### PR TITLE
#13070 - Add export (basic and detailed) and import

### LINKED ISSUE #13070: Add an Export (basic & detailed) and Import to Self Reporting [3]
### Feature Description

With the new Module Self-Reporting the users with the needed rights can also export Self-reported messages and import self-reported messags.
The export can be done as basic or detailed export

### Added Value/Benefit

As it's currently the only way how self-reported messages can enter the system, the import is essential.
The export is a basic functionality which we offer for every entity

### Acceptance Criteria

- [x] Import is possible, an import template can be provided, error handling for failed imports is available (as we do it for other imports
- [x] Basic Export is possible 
The basic export will contain all the fields which are displayed in the grid:
- Uuid
- Report date
- Type
- Disease
- First name
- Last name
- Birth date
- Sex
- District
- Address
- Phone number
- Email
- Responsible user
- Processing status
- Reason for deletion ( only when the "Deleted self reports" are selected from the grid)

- [x] Detailed Export is possible
  The detailed export will contain the following fields:
   - Uuid
   - Type
   - Report date
   - Case reference
   - Disease
   - Disease details
   - Disease Variant
   - Disease variant details
   - Firstname
   - Lastname
   - Sex
   - District
   - Street
   - House number
   - Postal Code
   - City
   - Birth date
   - National health id
   - Email
   - Phone number
   - Date of test
   - Date of symptoms
   - Workplace
   - Date of workplace
   - Isolation date
   - Contact date
   - Comment
   - Responsible user
   - Investigation status
   - Processing status
   - Deletion reason
   - Other deletion reason
 
- [x] Add user rights for Import/Export

### Implementation Details

Here is the structure of the file received when a self-report is made
 
The import template will contain the following fields: 
   - Type
   - Report date
   - Case reference
   - Disease
   - Disease details
   - Disease Variant
   - Disease variant details
   - Firstname
   - Lastname
   - Sex
   - Day of birth
   - Month of birth
   - Year of birth
   - National health id
   - Email
   - Phone Number
   - LocationDto address fields:
             - Continent
             - Subcontinent
             - Country
             - Region
             - District
             - Community
             - Community contact person
             - City
             - Area type
             - GPS latitude
             - GPS longitude
             - GPS accuracy in m
             - Postal code
             - Street
             - House number
             - Additional information
             - Address type
             - Address name/description
             - Facility type
             - Facility
             - Facility name & description
             - Contact person first name
             - Contact person last name
             - Contact person phone number
  - Date of test
  - Date of symptoms
  - Workplaces
  - Date workplace
  - Date of isolation
  - Date of contact
  -  Comment
 
The file is processed by an ETL job in the "Care+" section.

### Mockups

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by carina29**
Fixes #13070

### CODE REVIEW

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13103

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13103

File: sormas-backend/src/main/java/de/symeda/sormas/backend/importexport/ImportFacadeEjb.java [hunk 1]
```diff
@@ -529,6 +523,34 @@ public void generateEnvironmentImportTemplateFile(List<FeatureConfigurationDto>
...
+		char separator = configFacade.getCsvSeparator();
+		List<ImportColumn> importColumns = new ArrayList<>();
+
+		appendListOfFields(
+			importColumns,
+			SelfReportDto.class,
+			"",
+			separator,
+			featureConfigurations,
+			SelfReportDto.class.getDeclaredField("investigationStatus"),
```
Comment (leventegal-she) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/importexport/ImportFacadeEjb.java:540]: reflection is slow so here please use strings only, and use constants from the dto class

File: sormas-backend/src/main/java/de/symeda/sormas/backend/importexport/ImportFacadeEjb.java [hunk 2]
```diff
@@ -529,6 +523,34 @@ public void generateEnvironmentImportTemplateFile(List<FeatureConfigurationDto>
...
+			SelfReportDto.class.getDeclaredField("processingStatus"),
+			SelfReportDto.class.getDeclaredField("deleted"),
+			SelfReportDto.class.getDeclaredField("deletionReason"),
+			SelfReportDto.class.getDeclaredField("otherDeletionReason"),
+			SelfReportDto.class.getDeclaredField("INVESTIGATION_STATUS"),
```
Comment (leventegal-she) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/importexport/ImportFacadeEjb.java:545]: static fields are skipped in the `appendListOfFields` method

File: sormas-backend/src/main/java/de/symeda/sormas/backend/selfreport/SelfReportImportFacadeEjb.java [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+		String[][] entityPropertyPaths,
+		boolean ignoreEmptyEntries) {
+
+		// Check whether the new line has the same length as the header line
+		if (values.length > entityProperties.length) {
+			return ImportLineResultDto.errorResult(I18nProperties.getValidationError(Validations.importLineTooLong));
+		}
+
+		SelfReportDto selfReport;
+		if (values[0].isBlank() || !EnumUtils.isValidEnum(SelfReportType.class, values[0])) {
```
Comment (leventegal-she) [inline sormas-backend/src/main/java/de/symeda/sormas/backend/selfreport/SelfReportImportFacadeEjb.java:77]: having the type in the first column is not a must, there is no need to check it just do a build with null all the time

**Review by leventegal-she — CHANGES_REQUESTED** [review]

**Review by leventegal-she — CHANGES_REQUESTED** [review]

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13103

**Review by leventegal-she — APPROVED** [review]

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=13103

**Pull request merged**